src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java

```diff
@@ -314,15 +320,24 @@ private static LayoutPath derefPath(MemoryLayout layout, MethodHandle handle, La
 
     // Helper methods
 
-    private void check(Class<?> layoutClass, String msg) {
+    private <T extends MemoryLayout> T requireLayoutType(Class<T> layoutClass) {
         if (!layoutClass.isAssignableFrom(layout.getClass())) {
-            throw badLayoutPath(msg);
+            var name = layoutClass.getSimpleName();
+            var type = name
+                    // Take what is before "Layout"
+                    .substring(0, name.indexOf("Layout"))
+                    .toLowerCase(Locale.ROOT);
```

Comment:
I'd much prefer the simplicity of just passing in the name of the type as a `String` parameter.

Also, it seems like this wouldn't work for `AddressLayout` ?: `"attempting to select a address element from a non-address layout ..."`

Maybe a couple of simple `checkSequenceLayout`, `checkGroupLayout` would be better to avoid the existing duplication. I kinda feel like the new solution ends up introducing more complexity than it resolves.